File: MATTHEW/leetcode/AIcoding/5.Qwen-3/cpp_solutions/2369-check-if-there-is-a-valid-partition-for-the-array.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    bool validPartition(vector<int>& nums) {
        int n = nums.size();
        vector<bool> dp(n + 1, false);
        dp[0] = true;

        for (int i = 3; i <= n; ++i) {
            if (dp[i - 3] && nums[i - 1] == nums[i - 2] && nums[i - 2] == nums[i - 3]) {
                dp[i] = true;
            } else if (dp[i - 2] && nums[i - 1] == nums[i - 2]) {
                dp[i] = true;
            } else if (dp[i - 1] && nums[i - 1] == nums[i - 2] + 1 && nums[i - 2] == nums[i - 3] + 1) {
                dp[i] = true;
            }
        }

        return dp[n];
    }
};

File: MATTHEW/leetcode/AIcoding/4.GPT-4o/rust_solutions/2369-check-if-there-is-a-valid-partition-for-the-array.rs
impl Solution {
    pub fn valid_partition(nums: Vec<i32>) -> bool {
        let n = nums.len();
        let mut dp = vec![false; n + 1];
        dp[0] = true;

        for i in 2..=n {
            if nums[i - 1] == nums[i - 2] {
                dp[i] |= dp[i - 2];
            }
            if i > 2 && nums[i - 1] == nums[i - 2] && nums[i - 2] == nums[i - 3] {
                dp[i] |= dp[i - 3];
            }
            if i > 1 && nums[i - 1] == nums[i - 2] + 1 && nums[i - 2] == nums[i - 3] + 1 {
                dp[i] |= dp[i - 2];
            }
        }

        dp[n]
    }
}